source/credentials_provider_sts_web_identity.c

```diff
@@ -0,0 +1,1112 @@
+
+    if (aws_xml_parser_parse(&xml_parser, s_on_creds_node_encountered_fn, credentials)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "Failed to parse xml response for sts web identity with error %s",
+            aws_error_debug_str(aws_last_error()));
+        aws_xml_parser_clean_up(&xml_parser);
+        aws_credentials_destroy(credentials);
+        return NULL;
+    }
+
+    aws_xml_parser_clean_up(&xml_parser);
+    if (!credentials->access_key_id || !credentials->secret_access_key) {
+        aws_credentials_destroy(credentials);
+        return NULL;
+    }
+    return credentials;
+}
+
+/*
+ * No matter the result, this always gets called assuming that esc_user_data is successfully allocated
+ */
+static void s_sts_web_identity_finalize_get_credentials_query(
+    struct aws_credentials_provider_sts_web_identity_user_data *sts_web_identity_user_data) {
+    /* Try to build credentials from whatever, if anything, was in the result */
+    struct aws_credentials *credentials = NULL;
+    if (sts_web_identity_user_data->status_code == AWS_HTTP_STATUS_CODE_200_OK) {
+        credentials = s_sts_web_identity_parse_credentials_from_response(
+            sts_web_identity_user_data->allocator, &sts_web_identity_user_data->response);
+    }
+
+    if (credentials != NULL) {
+        AWS_LOGF_INFO(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p) STS_WEB_IDENTITY credentials provider successfully queried credentials",
+            (void *)sts_web_identity_user_data->sts_web_identity_provider);
+    } else {
+        AWS_LOGF_WARN(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p) STS_WEB_IDENTITY credentials provider failed to query credentials",
+            (void *)sts_web_identity_user_data->sts_web_identity_provider);
+    }
+
+    /* pass the credentials back */
+    sts_web_identity_user_data->original_callback(credentials, sts_web_identity_user_data->original_user_data);
+
+    /* clean up */
+    s_aws_credentials_provider_sts_web_identity_user_data_destroy(sts_web_identity_user_data);
+    aws_credentials_destroy(credentials);
+}
+
+static int s_sts_web_identity_on_incoming_body_fn(
+    struct aws_http_stream *stream,
+    const struct aws_byte_cursor *data,
+    void *user_data) {
+
+    (void)stream;
+
+    struct aws_credentials_provider_sts_web_identity_user_data *sts_web_identity_user_data = user_data;
+    struct aws_credentials_provider_sts_web_identity_impl *impl =
+        sts_web_identity_user_data->sts_web_identity_provider->impl;
+
+    AWS_LOGF_TRACE(
+        AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+        "(id=%p) STS_WEB_IDENTITY credentials provider received %zu response bytes",
+        (void *)sts_web_identity_user_data->sts_web_identity_provider,
+        data->len);
+
+    if (data->len + sts_web_identity_user_data->response.len > STS_WEB_IDENTITY_RESPONSE_SIZE_LIMIT) {
```

Comment:
since we're using dynamic buffer append apis here, why are we checking this?

Reply:
Followed the style used in IMDS, I think we do this as a safety concern, so as to discard unexpected response (which exceeded the expected length.)